---
title: "Alignment-Free-Classification"
author: "Micah Thornton"
date: "5/5/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Alignment Free Distances

When comparing genetic or proteomic sequence among one pair or more of organisms
it becomes necessary in most cases to align the sequences to one another such 
that there is some consistency in the symbols being compared.  This procedure 
usually known as multiple sequence alignment (MSA) is generally an expensive 
operation computationally and requires quite some time to complete.  In this 
study we will examine nearly 1,400 viromes (1,322) of the *SARS-CoV-2* genome that were 
downloaded in May 2020 from the GISAID Initiative (an initiative which curates 
and maintains not only viromes from *SARS-CoV-2*, but also other pathogenic 
viruses such as *Influenza*). The YinGenomicDFTDistances package provides the 
dataset that will be investigated as a collection of headers and the associated
*FASTA* formatted sequences with which that header is associated. 

```{r loadingData}
library(YinGenomicDFTDistances); 
viralData <- YinGenomicDFTDistances::sarscvmay;
str(viralData)
```

# Parsing the Headers 
As we can see, the data contained in the header field of the packaged dataset 
is not in a usable format, but may be parsed using regular expressions to 
produce a set of valid submitting laboratory locations.  This parsing may be 
executed in a variety of different ways, but here we use string pattern matching
with regular expressions. 

```{r parseHeaders}
viralData$Location <- unlist(lapply(strsplit(as.character(viralData$Header),"/"), function(x){x[2]}));
for (inc in c('tiger','env')){
  viralData[viralData$Location == inc, ]$Location <- unlist(lapply(strsplit(as.character(viralData[viralData$Location == inc,]$Header),"/"), function(x){x[3]}))
}
viralData$Region <- rep("blank",nrow(viralData)); 
if (nrow(viralData[viralData$Location %in% c("Algeria", "Egypt", "South Africa",  "DRC",  "Gambia", "Senegal"),]) > 0){
viralData[viralData$Location %in% c("Algeria", "Egypt", "South Africa",  "DRC",  "Gambia", "Senegal"),]$Region <- "Africa"
}
if (nrow(viralData[viralData$Location %in% c("Beijing" ,"Chongqing" ,"Fujian" ,"Guangdong" ,"Hangzhou" , "Hong Kong" ,"Jiangsu" ,"Jiangxi" ,"Jingzhou" ,"Shandong" ,"Shenzhen" , "Sichuan" , "Taiwan", "Tianmen" ,"Wuhan", "Yunnan" , "Zhejiang" ,"Lishui"  ,"Japan"  ,"Guangzhou"),]) > 0){
viralData[viralData$Location %in% c("Beijing" ,"Chongqing" ,"Fujian" ,"Guangdong" ,"Hangzhou" , "Hong Kong" ,"Jiangsu" ,"Jiangxi" ,"Jingzhou" ,"Shandong" ,"Shenzhen" , "Sichuan" , "Taiwan", "Tianmen" ,"Wuhan", "Yunnan" , "Zhejiang" ,"Lishui"  ,"Japan"  ,"Guangzhou"),]$Region <- "East Asia"
}
if (nrow(viralData[viralData$Location %in% c( "Brazil" ,"Chile" ,"Colombia" ,"Costa Rica" ,"Uruguay"),]) > 0){
viralData[viralData$Location %in% c( "Brazil" ,"Chile" ,"Colombia" ,"Costa Rica" ,"Uruguay"),]$Region <- "South America"
}
if (nrow(viralData[viralData$Location %in% c("Austria", "Belgium", "Czech Republic", "Denmark", "Finland", "France" , "Georgia", "Germany" , "Greece" , "Spain", "Sweden" , "Hungary" , "Portugal" , "Poland", "Russia", "Romania" , "Slovakia" , "Italy"),]) > 0){
viralData[viralData$Location %in% c("Austria", "Belgium", "Czech Republic", "Denmark", "Finland", "France" , "Georgia", "Germany" , "Greece" , "Spain", "Sweden" , "Hungary" , "Portugal" , "Poland", "Russia", "Romania" , "Slovakia" , "Italy"),]$Region <- "Europe"
}
if (nrow(viralData[viralData$Location %in% c( "Bangladesh" , "Cambodia" , "India" , "Nepal" , "Sri Lanka" , "Vietnam" , "Thailand"),]) > 0){
viralData[viralData$Location %in% c( "Bangladesh" , "Cambodia" , "India" , "Nepal" , "Sri Lanka" , "Vietnam" , "Thailand"),]$Region <- "West Asia"
}
if (nrow(viralData[viralData$Location %in% c( "Australia", "New Zealand" , "Indonesia" , "Malaysia"),]) > 0){
viralData[viralData$Location %in% c( "Australia", "New Zealand" , "Indonesia" , "Malaysia"),]$Region <- "Oceania"
}
if (nrow(viralData[viralData$Location %in% c("Turkey" , "Saudi Arabia" , "Kazakhstan" ,"Iran", "Israel" , "Kuwait"),]) > 0){
viralData[viralData$Location %in% c("Turkey" , "Saudi Arabia" , "Kazakhstan" ,"Iran", "Israel" , "Kuwait"),]$Region <- "Middle East"
}
if (nrow(viralData[viralData$Location %in% c("USA", "Puerto Rico", "Guam", "Mexico"),]) > 0){
viralData[viralData$Location %in% c("USA", "Puerto Rico", "Guam", "Mexico"),]$Region <- "North America"
}
if (nrow(viralData[viralData$Location %in% c("blank"),]) > 0){
viralData[viralData$Location %in% c("blank"),]$Region <- viralData[viralData$Location %in% c("blank"),]$Location
}
table(viralData$Region)
```
# Computing the Alignment-Free Distance metrics with YinDFTDistances Package 

The YinGenomicDFTDistances Package provides two alignment-free methods for 
determining the distance between different sequences of the same type.  The 
reason why these are known as 'alignment-free' methods is simply because they 
do not require the sequences to be aligned to one another using a Multiple 
Sequence Alignment (MSA) tool such as clustalW or Clustal Omega prior to the 
computation of the metric.  The two such methods provided by the package are 
(I) the First-Five K-mer Frequency Distance, and (II) The DFT spectra distances. 

The 
## The First-Five K-mer Frequency Distance 